ouroboros-network/src/Ouroboros/Network/Connections/Types.hs

```diff
@@ -0,0 +1,178 @@
+mapDecision frej facc decision = case decision of
+  Rejected rej -> Rejected (frej rej)
+  Accepted acc -> Accepted (facc acc)
+
+-- | Useful type with kind `Provenance -> Type` to express that rejection is
+-- not possible.
+data CannotReject (provenance :: Provenance)
+
+-- | Useful for concurrent connections manager which have only one request
+-- type that carries no information.
+data UnitRequest (p :: Provenance) where
+  UnitRequest :: UnitRequest p
+
+data LocalOnlyRequest (p :: Provenance) where
+  LocalOnlyRequest :: LocalOnlyRequest Local
+
+data RemoteOnlyRequest (p :: Provenance) where
+  RemoteOnlyRequest :: RemoteOnlyRequest Remote
+
+-- | Description of a resource: something that can be acquired in the bracket
+-- pattern, or something that already exists but can be closed.
+--
+-- This is indexed by `Provenance` because incoming and outgoing resources
+-- have different contracts: incoming ones already exist but can be closed;
+-- outgoing ones _may_ be created but may also already exist.
+data Resource provenance m r where
+  -- | An existing resource, with a close action.
+  -- Corresponds to remotely-initiated, incoming connections.
+  Existing ::   r -> m ()        -> Resource Remote m r
+  -- | A new resource, with create and close actions (a bracket).
+  -- Corresponds to locally-initiated, outgoing connections.
+  New      :: m r -> (r -> m ()) -> Resource Local m r
+
+-- | Connections identified by `id`, carried by `resource`, supported by `m`.
+-- `resource` might be a file handle / socket, STM queue, etc. anything that
+-- can transport data in `m`.
+--
+-- For `Existing` resources, if the decision is `Rejected`, or if the call
+-- throws an exception (`m` is IO-capable), then the caller is responsible for
+-- closing the resource. Otherwise, the `Connections` term must ensure it is
+-- closed appropriately.
+--
+-- For `New` resources, the caller does not create the resource, so the
+-- `Connections` term is responsible for creating and closing it appropriately.
+--
+-- Note on exceptions: in case `m` can throw exceptions, there's some
+-- ambiguity about when it's a good idea to do so, because it's not clear
+-- whether an exception should be caught and used to give a `Rejected` value,
+-- or simply ignore/re-thrown.
+-- If the resource acquisition of a `New` connection throws an exception, that
+-- should be re-thrown. This is consistent with bracketing of the acquire and
+-- release fields of the `New` constructor, which also re-throws the exception.
+--
+-- TODO may need a way to check / be notified when there is an incoming
+-- connection at a given identifier.
+-- Could just put that on the concurrent connections: an `STM (Map id handle)`.
+-- When you get the handle back, the underlying connection and its handler
+-- thread may of course have died, but that's always a possibility even if
+-- you `include`. 
+newtype Connections id resource request reject accept m = Connections
```

Comment:
The type name is misleading, what about changing it to `ConnectionDecision` or `IncludeConnection`.

Reply:
I take my comment back.  But I would add a comment that the closure of the `Connection` contains a mutable state of all the connections, which justifies the name.